### PR TITLE
raylib-rs 6.0 — release prep checkpoint

### PR DESCRIPTION
**Pull request by Dacode45**
## raylib-rs 6.0 — release prep checkpoint

This PR bundles **WS1–WS8** of the raylib 5.x → 6.0 upgrade for owner review.
It is a **checkpoint, not a release** — no crates.io publish, no canonical
merge, no `v6.0.0` tag. After this PR merges, the fork's `unstable` HEAD is
the snapshot the owner evaluates before WS9 (showcase rewrite) and any
further work they choose to add.

### Workstream summary

- **WS0** ✅ Baseline + roadmap — `docs/superpowers/specs/2026-05-25-raylib-rs-6.0-roadmap-design.md`.
- **WS1** ✅ raylib-sys at 6.0 parity (3-OS CI green) — `docs/superpowers/notes/ws1-config-reconcile.md`.
- **WS2a** ✅ Native math types (bindgen-generated `Vector2/3/4`, `Matrix`, distinct `Quaternion`).
- **WS2b** ✅ Math-ecosystem adapters opt-in only; `mint`/`glam`/`serde` no longer default.
- **WS3** ✅ Safe crate green vs 6.0; `ModelAnimations` RAII; Tier-1 tests — `docs/superpowers/notes/ws3-complete.md`.
- **WS4** ✅ Software-renderer + headless render-test harness — `docs/superpowers/notes/ws4b-complete.md`.
- **WS5** ✅ raygui 6.0 parity (57/57 fns) + safe `rlgl` immediate-mode module — `docs/superpowers/notes/ws5-complete.md`.
- **WS6** ✅ Platform matrix + layered CI/CD + quality hard-gates + MSRV 1.85 — `docs/superpowers/notes/ws6b-complete.md`.
- **WS7** ✅ 28-chapter mdBook + ~25 rustdoc enrichments + CHANGELOG 6.0.0 entry + backlog #284/#273 folded — `docs/superpowers/notes/ws7-complete.md`.
- **WS8** ✅ Version bump + draft release workflows (`release-sys.yml`, `release-safe.yml` — workflow_dispatch-only, dry_run-gated) + `act` validation + release-hygiene fold-ins (Node.js 24 actions, `paste` advisory acceptance, `samples/` removal) — `docs/superpowers/notes/ws8-complete.md` (added in this PR after merge).

### What's deliberately NOT in this PR

- **No publish to crates.io.** `release-sys.yml` and `release-safe.yml` are authored but never auto-triggered. Real publish is the future final-release step.
- **No CHANGELOG date flip.** `## 6.0.0 (unreleased)` heading stays; the date is set at real publish time.
- **No `v6.0.0` tag.** Tagging is part of the final-release step.
- **No PR to `raylib-rs/raylib-rs`.** The canonical merge happens only after WS9 and any added work.

### Review focus areas

- Per-WS commit grouping is preserved. Each WS lands as a sequence of small commits with `Co-Authored-By: Claude Opus 4.7` attribution. Backlog-PR cherry-picks (PR #284 LBreede, PR #273 AmityWilder, PR #272 AmityWilder, etc.) keep the original `--author` headers.
- Locked WS8 decisions are recorded in the spec §3.
- Tracked-deferred items (see WS8 spec §8 + each `wsN-complete.md`) carry forward.

### CI surface

All 5 existing workflows + 2 new release workflows present on this branch. The 5 auto-running ones are green at the PR HEAD. The 2 new release workflows are `workflow_dispatch`-only and never auto-trigger.

### How to merge

Select **"Create a merge commit"** in the GitHub UI (`--no-ff` semantics). Do **not** select squash or rebase-and-merge — those destroy the per-WS commit groupings and backlog-PR attribution headers.

### CODE REVIEW

File: raylib/src/rgui/controls.rs [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+        }
+        v.set_len(capacity);
+    }
+    let ptr = buffer.as_mut_ptr() as *mut c_char;
+    let res = call(ptr, capacity as i32);
+    // SAFETY: every byte in [0..capacity] was initialized above (and C only writes
+    // initialized bytes over it), so this shared byte view is sound.
+    let scanned = unsafe { std::slice::from_raw_parts(buffer.as_ptr(), capacity) };
+    let len = scanned.iter().position(|&b| b == 0).unwrap_or(capacity);
+    // SAFETY: `len <= capacity`; bytes [0..len] contain no NUL and are valid UTF-8
```
Comment (Dacode45) [inline raylib/src/rgui/controls.rs]: I want a debuh_assert here

File: raylib/src/rlgl/immediate.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    /// Set the current texture coordinate.
+    #[inline]
+    pub fn texcoord2f(&mut self, x: f32, y: f32) {
+        unsafe { ffi::rlTexCoord2f(x, y) }
+    }
+
+    /// Set the current normal vector.
+    #[inline]
+    pub fn normal3f(&mut self, x: f32, y: f32, z: f32) {
```
Comment (Dacode45) [inline raylib/src/rlgl/immediate.rs]: Double check that no rlgl functions are missing

**Comment by Dacode45** [comment]
## Review feedback addressed — commit `58c5224`

Thanks for the review. Triaged your 10 inline comments per your direction (fix if small, future workstream if not). 4 fixed in `58c5224`, 1 already covered by an existing test, 5 logged as future workstreams. Full breakdown in `docs/superpowers/notes/ws8e-checkpoint-review-feedback.md` on this branch.

### Fixed in `58c5224`

| # | File | Comment | Fix |
|---|------|---------|-----|
| 1 | `.github/workflows/book.yml` | Should use an apt cache here | Added `awalsh128/cache-apt-pkgs-action@v1` (cache key `6.0-book-v1`) |
| 2 | `.github/workflows/book.yml` | should use rust caching here | Added `Swatinem/rust-cache@v2` (`shared-key: book`) between toolchain + `cargo build` |
| 3 | `.github/workflows/release-sys.yml` | Validate publish doesn't include native raylib examples | Added a `cargo package --list` grep that fails the build if any `(raylib/)?(examples\|projects\|templates)/` paths leak past the `Cargo.toml` exclude list |
| 9 | `raylib/src/rgui/controls.rs` | I want a debug_assert here | Added `debug_assert!(min_value <= max_value, ...)` with formatted message to `gui_spinner`, `gui_value_box`, `gui_slider`, `gui_slider_bar`, `gui_progress_bar` |

### Already verified by existing test

| # | File | Comment | Resolution |
|---|------|---------|-----------|
| 6 | `raylib-sys/src/glam_conv.rs` | Double check glam matrix index ordering | Covered by `raylib-sys/tests/conversions.rs:251` (`glam_matrix_translation_matches_semantically`) — asserts `Matrix::translate(1,2,3)` == `glam::Mat4::from_translation(vec3(1,2,3))` via `abs_diff_eq(1e-5)`. If column-major ordering disagreed the matrices would differ at the translation column; the test forces correctness. Module-level docs at `glam_conv.rs:8-19` spell out the mapping. |

### Logged as future workstreams (post-WS8 checkpoint)

| # | File | Comment | Workstream |
|---|------|---------|------------|
| 4 | `.github/workflows/test.yml` | Validate `nobuild` mode + bring-your-own raylib | "Nobuild-mode CI matrix" — fetches the prebuilt libs from <https://github.com/raysan5/raylib/releases/tag/6.0>, sets the bindgen-location env, adds a new matrix dimension. Non-trivial; post-release. |
| 5 | `raylib-sys/src/color.rs` | `impl From<&Color> for Color` redundant if Clone is derived | "Color/Vector conversion ergonomics audit" — `Color: Copy + Clone` is already derived; the explicit `From<&Color>` exists for `Into<Color>` constraints on `&Color` receivers. Removing requires auditing all callsites + the analogous impls on Vector2/3/4 / Rectangle. Focused refactor workstream post-release. |
| 7 | `raylib/src/core/callbacks/audio_stream_callback.rs` | Use `thiserror` for all error types | "thiserror migration" — `thiserror = "2.0.12"` is already in `[dependencies]`. Migration converts remaining ad-hoc error types crate-wide. Needs its own brainstorm to decide on the canonical error hierarchy. |
| 8 | `raylib/src/core/databuf.rs` | More tests for edge cases, software renderer, valgrind | "DataBuf testing + memory profiling" — per your "more tests can be a future workstream" reply. Uses the existing software-renderer Tier-2 harness + valgrind/Miri for the `ManuallyDrop<Box<[T]>>` wrappers. Likely an umbrella workstream covering DataBuf, Mesh accessors, FilePathList, ImageBuf. |
| 10 | `raylib/src/rlgl/immediate.rs` | Double-check no rlgl functions are missing | "rlgl safe-module coverage audit" — diff the safe `rlgl` public surface against the full `ffi::rl*` function list; gaps either get wrapped, documented as escape-hatch FFI, or recorded as future work. Pairs naturally with the testing workstream. |

### Post-release direction noted

You mentioned wanting to start a **bevy-raylib** crate post-release. Saved to project memory so it surfaces in future sessions; sketched into the workstream ordering at the bottom of `ws8e-checkpoint-review-feedback.md`.

### Suggested workstream ordering after this checkpoint

1. **WS9** — showcase rewrite + Pages deploy (roadmap finale).
2. **Final-release** — publish to crates.io + canonical merge + tag + GitHub release.
3. **bevy-raylib** — new crate depending on the published `raylib 6.0.0`.
4. **DataBuf + Mesh + general testing umbrella** (covers comments 8 + 10).
5. **thiserror migration** (comment 7).
6. **Nobuild-mode CI matrix** (comment 4).
7. **Color/Vector conversion ergonomics audit** (comment 5).
8. **Symmetric apt+rust-cache adoption** across `check.yml` / `test.yml` / `web.yml` / `sanitizers.yml` (extending what comments 1+2 did for `book.yml`).

Items 4-8 are flexible — reorder when each one actually starts based on what's most painful at the time.

Once you submit the review and approve, I'll execute WS8e Tasks 5-9 (merge with "Create a merge commit", flip the `CLAUDE.md` status to ✅ WS8, write the `ws8-complete.md` done-note, dispatch the final overall code review).
